README.md
# Bilibili Downloader

无聊写的一个程序，基于 PyQt5 。

此程序只可以下载 B站 上由 UP主 上传的视频 (即链接为 www.bilibili.com/video/BVxxxxxxxxxx 格式的)。

**警告:** 此程序不支持在 Windows 以外的平台运行。

本人一般不会对其进行维护，除非出现重大错误 ~~(或者作者闲的无聊)~~。

## 使用

您可以在 Releases 获得每次更新的 Windows 可执行文件。

也可以直接运行 main.py (**对于 macOS / Linux 用户, 这是最好的**) 。

如果您要直接运行 main.py 来使用此程序，请先在命令行里输入以下命令:

    对于 macOS / Linux 用户:
    python3 -m pip install -r requirements.txt
    对于 Windows 用户:
    pip3 install -r reqirements.txt

每次下载视频后都会把下载的视频名和视频地址写入程序所在目录下的 "DownloadHistory.txt" 文件。(不会记录下载时间)

## 关于 FFMPEG

此程序已内置 [FFMPEG](https://ffmpeg.org) (版本 N-102989-g7f6d20931b-20210717, 来自 [BtbN 的 Github 构建仓库](https://github.com/BtbN/FFmpeg-Builds))。

此程序除 FFMPEG 程序遵守 LGPL 许可证外, 其余代码以 MIT 协议和 "特此声明" 部分的条款 授权。

FFMPEG (C) 2000-2021 the FFmpeg developers.

## 特此声明

若您使用本程序，则代表您同意以下条款 和 MIT 协议，否则您无权使用此软件:

***此软件禁止用于商业用途，使用此软件引发的任何后果与作者无关，作者不承担责任。***


CHANGES.md
# Bilibili Downloader 更新日志

## v1.2.0

* 大量重构代码 (重构了 95% 的代码)，优化用户体验。

* 对只在程序内部调用的所有方法进行调整，使其只能由内部调用。

* 优化并格式化核心程序。

* 增加了选择视频格式的选择框。

* 增加了将视频转为 MP3 的选项。

* 增加了在转换 MP4 时显示百分比进度。 (参考自 [lvshunxia 的 CSDN 博客](https://blog.csdn.net/lvshunxia/article/details/104972424))

* 修复了下载线程 (QThread) 在获取文件阶段可能产生错误但无法被处理的问题。

* 修复了当用户下载 360P 视频时仍会用 FFMPEG 进行转码的问题 (因为 360P 的原始视频文件已经是 MP4 格式)。

* 修复在找不到目标清晰度时，弹出的提示框的最高清晰度为数字的问题。

## v1.1.0

* 增加 ffmpeg 自动转 MP4 支持 (已测试)

* 更智能的提示 (使用 PyQt 信号-槽 机制)

## v1.0.0

* 最初版本。


main.py
import os
import subprocess
import sys

from PyQt5.QtCore import QThread, pyqtSignal, pyqtSlot
from PySide2.QtGui import QCloseEvent

from functions import *
from ui import *

headers = {
    'Referer': 'https://www.bilibili.com/',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36'
}


class DownloadThread(QThread):  # 继承QThread
    # 注册信号
    downloadOKSignal = pyqtSignal(str)
    formattingSignal = pyqtSignal(str)
    ffmpegNotFoundSignal = pyqtSignal(str)
    fileAlreadyExistsSignal = pyqtSignal(str)

    def __init__(self, vname, bv, flv_url, headers, page_num, is_will_convert, save_dir, no_delete_original_video, quality, videoNameLabel, is_will_delete_exists_file, videoFormatType):  # 从前端界面中传递参数到这个任务后台
        super().__init__()
        self.vname = vname
        self.bv = bv
        self.flv_url = flv_url
        self.headers = headers
        self.page_num = page_num
        self.is_will_convert = is_will_convert
        self.savedir = save_dir
        self.no_delete_original_video = no_delete_original_video
        self.quality = quality
        self.videoNameLabel = videoNameLabel
        self.is_will_delete_exists_file = is_will_delete_exists_file
        self.videoFormatType = videoFormatType

    def run(self):  # 重写run  比较耗时的后台任务可以在这里运行
        # 尝试下载, 如果失败发送下载完成信号(但参数为 None, 会使函数产生下载失败提示框) 并终止线程
        try:
            gettedFile = os.path.abspath(get_flv(
                self.vname, self.bv, self.flv_url, self.headers, self.page_num))  # 下载 FLV 文件
        except:
            self.downloadOKSignal.emit(None)
            return

        if (self.videoFormatType == 'flv') or (not self.is_will_delete_exists_file):  # 如果目标格式是 FLV
            self.downloadOKSignal.emit(gettedFile)  # 发送完成信号
            return

        outputfile = gettedFile.replace(
            '.flv', '.' + self.videoFormatType)  # 准备 转换 的 目标文件名

        # region 如果用户下载的是 360P 流畅 的视频文件 (格式已经为 MP4, 无需转换) 或者 用户请求不转 MP4, 发送下载完成信号, 终止线程
        if not self.is_will_convert:
            if self.quality == 16:  # 如果下载的是 360P 的视频文件
                # 获取的视频文件默认为 MP4 格式, 但保存的文件格式为 FLV, 要进行重命名
                os.rename(gettedFile, outputfile)
                self.downloadOKSignal.emit(gettedFile)  # 发送完成信号
                return
        # endregion

        # region 如果 ffmpeg.exe 不存在, 发送"FFMPEG未找到"信号, 终止线程
        if not os.path.isfile('ffmpeg.exe'):
            self.ffmpegNotFoundSignal.emit(gettedFile)
            return
        # endregion

        # region 检查是否要求删除已存在文件
        if self.is_will_delete_exists_file and os.path.exists(outputfile):
            os.remove(outputfile)
        # endregion

        self.formattingSignal.emit(gettedFile)  # 下载 FLV 完成, 发送转换信号

        # region 开始转 MP4
        def get_seconds(time):
            h = int(time[0:2])
            m = int(time[3:5])
            s = int(time[6:8])
            ms = int(time[9:12])
            ts = (h * 60 * 60) + (m * 60) + s + (ms / 1000)
            return ts

        process = subprocess.Popen('ffmpeg.exe -i "{inputfile}" "{outputfile}"'.format(
            inputfile=gettedFile, outputfile=outputfile), stderr=subprocess.PIPE, bufsize=0, universal_newlines=True, shell=True, encoding="utf-8")

        duration = None
        while process.poll() is None:
            line = process.stderr.readline().strip()
            if line:
                duration_res = re.search(r'Duration: (?P<duration>\S+)', line)
                if duration_res is not None:
                    duration = duration_res.groupdict()['duration']
                    duration = re.sub(r',', '', duration)

                result = re.search(r'time=(?P<time>\S+)', line)
                if result is not None and duration is not None:
                    elapsed_time = result.groupdict()['time']

                    currentTime = get_seconds(elapsed_time)
                    allTime = get_seconds(duration)

                    progress = currentTime * 100/allTime
                    self.videoNameLabel.setText(
                        'FLV 获取成功, 正在转 ' + self.videoFormatType.upper() + ' , 请耐心等待 ... ' + str(int(progress)) + '%')
        # endregion

        if not self.no_delete_original_video:  # 如果用户没有选择保留原始视频文件
            os.remove(gettedFile)  # 删除不再需要的 FLV 文件

        self.downloadOKSignal.emit(outputfile)  # 转 MP4 完成, 发送信号



class DownloaderUI(QMainWindow, Ui_MainWindow):
    def __init__(self, parent=None) -> None:
        super().__init__(parent)
        # region 保存位置初始化
        self.savedir = os.path.abspath('./Downloads')
        if not os.path.exists(self.savedir):
            os.mkdir(self.savedir)
        # endregion

        # region 其他初始化
        self.currentBV = None  # 初始化 "当前 BV" 属性
        self.w = QMainWindow()  # 创建主窗口
        self.setupUi(self.w)  # 初始化窗口组件
        self.w.setWindowFlags(Qt.WindowMinimizeButtonHint |
                              Qt.MSWindowsFixedSizeDialogHint | Qt.WindowCloseButtonHint)  # 设置窗口属性
        # endregion

        # region 信号绑定
        self.lineEdit.editingFinished.connect(self.__changename)
        self.download.clicked.connect(self.__download)
        self.saveAtButton.clicked.connect(self.__asksaveat)
        # endregion

    @pyqtSlot(int)
    def __changename(self):
        if self.lineEdit.text() == "":
            return
        try:
            self.currentBV = bvre(self.lineEdit.text())
        except:
            QMessageBox.critical(self, '错误', '无效的 BV 号, 请检查您的输入!')
            return
        else:
            try:
                self.videoName.setText(
                    '视频名称: ' + name(self.currentBV, headers))
            except:
                QMessageBox.critical(self, '错误', '无法获取视频名称, 请检查您的网络环境!')
                return

    @pyqtSlot(int)
    def __download(self):
        # 禁用一些控件
        self.__controlWidgets(False)

        videoFormat = self.videoFormatType.currentText().lower() # 视频格式

        # region 是否在 360P 下使用 FLV 格式
        if self.quality.currentText() == '360P' and videoFormat == 'flv':
            QMessageBox.critical(self, '错误!', '360P 清晰度的视频不支持 FLV 格式!')
            self.__controlWidgets(True)
            return
        # endregion

        # region 获取一些数据 (CID)
        page_num = int(self.videoPage.text())  # 页码
        try:
            cid = get_cid(self.currentBV, page_num)
        except CIDNotFoundException:  # CID 不存在
            QMessageBox.critical(
                self, '错误!', '您要下载的视频并没有第 ' + self.videoPage.text() + ' 页。')
            self.__controlWidgets(True)
        except:  # 其他网络错误
            QMessageBox.critical(self, '错误!', '获取视频 CID 失败, 请检查您的网络是否正常!')
            self.__controlWidgets(True)
            return
        # endregion

        vname = name(self.currentBV, headers)

        # region 获取 清晰度
        int_quality = qualityid(self.quality.currentText())
        text = play_url(cid, self.currentBV, headers, str(int_quality))
        accept_quality = text['data']['accept_quality']
        # endregion

        # region 检查 目标视频 是否支持选定的清晰度
        if not int_quality in accept_quality:  # 如果没有找到
            for x in text['data']['support_formats']:
                if max(accept_quality) == x['quality']:
                    max_quality = x['new_description']
            QMessageBox.critical(
                self, '下载失败!', '您要下载的视频不支持选择的清晰度!\n此视频支持的最佳画质: ' + max_quality)
            self.__controlWidgets(True)
            return
        # endregion

        # region 检查用户下载的视频的清晰度是否为 360P, 或者用户选择的视频格式不是 MP4, 如果是, 则在开始下载时为下载线程传入不转换 的参数。
        is_will_convert = True
        if int_quality == 16 or videoFormat == 'flv':
            is_will_convert = False
        # endregion

        # region 如果用户希望转视频格式为 其他 则检查文件是否已存在
        is_will_delete_exists_file = False
        testname = os.path.join(
            self.savedir, vname + ' - PAGE ' + self.videoPage.text() + '.' + videoFormat)
        if os.path.exists(testname):
            if QMessageBox.question(self, '提示!', testname + ' 已经存在! 是否覆盖?\n若不覆盖, 则只会下载原始视频文件。') == QMessageBox.Yes:
                is_will_delete_exists_file = True
        # endregion

        self.videoName.setText('正在下载...')  # 设置 下载提示文本

        # 准备 QThread
        self.DownloadQThread = DownloadThread(vname, self.currentBV, text['data']['durl'][0]['url'], headers,
                                              page_num, is_will_convert, self.savedir, self.saveOriginalVideo.isChecked(), int_quality, self.videoName, is_will_delete_exists_file, self.videoFormatType.currentText().lower())
        # region 下载线程的信号连接
        self.DownloadQThread.formattingSignal.connect(
            lambda: self.videoName.setText('FLV 获取成功, 正在转 ' + videoFormat.upper() + ' , 请耐心等待 ...'))
        self.DownloadQThread.ffmpegNotFoundSignal.connect(
            self.__ffmpegNotFound)
        self.DownloadQThread.downloadOKSignal.connect(
            self.__DownloadOK)  # 把任务完成的信号与任务完成后处理的槽函数绑定
        # endregion

        self.DownloadQThread.start()  # 开始下载

    def __controlWidgets(self, arg_bool: bool):
        self.lineEdit.setEnabled(arg_bool)
        self.download.setEnabled(arg_bool)
        self.saveAtButton.setEnabled(arg_bool)
        self.videoPage.setEnabled(arg_bool)
        self.quality.setEnabled(arg_bool)
        self.videoFormatType.setEnabled(arg_bool)
        self.saveOriginalVideo.setEnabled(arg_bool)

    def __ResetInputData(self):
        self.videoName.setText('输入完 BV 号后, 在输入框里按下回车即可在这里看到视频名称!')
        self.lineEdit.clear()
        self.__controlWidgets(True)

    def __DownloadOK(self, return_value):
        if os.path.isfile(return_value):
            QMessageBox.information(
                self, '下载完成!', '保存到: ' + return_value)
        else:
            QMessageBox.critical(self, '下载失败!', '下载失败! 请再试一次!')
        self.__ResetInputData()

    def __ffmpegNotFound(self, filename):
        QMessageBox.critical(
            self, '转 MP4 失败!', 'ffmpeg.exe 没有找到! 无法转 MP4 ...\n已下载的原始 FLV 文件: ' + filename)
        self.__ResetInputData()

    def __asksaveat(self):
        self.savedir = QFileDialog.getExistingDirectory(
            self, "请选择文件夹路径", os.getcwd())

    def show(self):
        self.w.show()


if __name__ == "__main__":
    if sys.platform == 'win32':
        import ctypes
        ctypes.windll.shell32.SetCurrentProcessExplicitAppUserModelID(
            "bilidownloader")
    app = QApplication(sys.argv)
    app.setStyle(QStyleFactory.create('Fusion'))
    ui = DownloaderUI()
    ui.show()
    return_code = app.exec_()
    # os.system('taskkill /f /im ffmpeg.exe')
    sys.exit(return_code)


icon_rc.py
# -*- coding: utf-8 -*-

# Resource object code
#
# Created by: The Resource Compiler for PyQt5 (Qt v5.15.2)
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore

qt_resource_data = b"\
\x00\x00\x01\x9c\
\x00\
\x00\x10\xbe\x78\x9c\xed\x98\x31\x4f\xc2\x40\x14\xc7\x9f\x29\xa1\
\x83\x09\x35\x31\x31\x34\x0c\x65\xf4\x5b\xe0\xaa\x5f\x82\x44\x67\
\xe1\x2b\xf8\x51\x74\xe3\x0b\xe8\xcc\x62\x1c\x1c\xd0\xa5\x33\x83\
\x29\x89\x2e\x4d\x60\xa8\x09\xc9\xf9\x1e\xbc\x8b\x67\x39\x7a\x3d\
\x68\x0b\x46\xaf\xf9\x91\x72\xbd\xfe\x7f\xc7\x5d\x48\x1f\x00\x1c\
\xe0\xd1\x6e\x03\xbd\xc2\xe0\x08\xe0\x04\x00\x4e\x11\xec\x82\x33\
\x58\xf6\x2f\x1a\x5e\x3b\x3e\x5c\x22\x9b\x10\xe2\x4f\x13\xde\x81\
\xa4\x8e\xf4\x91\x67\x64\x86\x88\x82\x99\x71\x76\x9f\x5d\xaa\xbb\
\x85\x8c\x4a\x70\xae\xe3\x85\x9d\xe4\x76\xf9\x7d\x55\x6e\xc9\x2b\
\xbb\x7b\x3b\x70\x4b\xc8\xfd\x94\xea\xbb\x47\x9a\xa9\xfd\x29\x82\
\x16\x67\xab\x2e\x72\x7f\xa6\xfa\xfc\x12\xdc\x92\x66\xca\x35\xd5\
\xac\x49\x59\x6e\x89\x69\x4f\xfe\xfd\xfb\xe5\xbf\x40\x26\xc8\x1b\
\x9f\x9b\xf2\x4d\xe3\x6d\xfd\x91\x72\x6d\x92\xc3\x6f\x1a\x6f\xeb\
\xb7\xdd\x1f\xdb\xbc\x7d\xf7\xab\xeb\x19\x59\xae\xbf\x6e\xbc\xad\
\xff\x9c\x73\x22\x3e\x37\xf9\x4d\xe3\x7f\xdb\xf7\xaf\x6c\x7f\x95\
\xcf\x1f\x3f\xe5\x4a\xc2\xd5\xe7\xef\x43\x49\x73\xf0\x39\x5b\x75\
\x91\xfb\x3a\xc7\x9e\x94\x05\xb9\xa9\x16\xac\xb2\xf6\x93\x50\xcd\
\xe7\xf2\xda\x54\x5d\x7f\x8e\xc2\xef\xfa\x53\x42\xeb\x20\xeb\x31\
\x5d\x5d\xb2\x2d\x53\xce\xee\x85\xab\xf5\xb7\x8e\x06\xf2\xa8\xc9\
\xa1\x3e\xaf\x80\xf1\x59\xd4\x37\xcc\xca\x9a\x43\x9e\xcf\x2c\xb9\
\x5c\x93\xd1\xd8\x62\xdd\xae\x2c\xfc\xdd\x0d\xdd\x59\x73\xe8\x5a\
\xdc\x5f\x43\x6e\x91\x77\x64\x60\xe9\x56\xe7\x40\xf7\x7e\x70\x56\
\x4d\x37\x6e\xd7\xbf\x7f\x05\xb7\x21\x80\x73\x43\x7f\x15\x70\x73\
\xb0\xcb\x15\x62\xec\x09\x11\xff\xc0\x8b\xf5\xf0\x75\xba\xc7\x59\
\xc4\x2d\x1b\x65\x52\xf6\xdc\x19\x76\x12\x77\x1c\x24\x5e\x1c\x24\
\x74\x74\xe6\x01\x79\xbf\x00\x7e\x98\x6c\x41\
"

qt_resource_name = b"\
\x00\x04\
\x00\x06\xfa\x5e\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\
\x00\x08\
\x0a\x61\x42\x7f\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\x00\x2e\x00\x69\x00\x63\x00\x6f\
"

qt_resource_struct_v1 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x0e\x00\x01\x00\x00\x00\x01\x00\x00\x00\x00\
"

qt_resource_struct_v2 = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x00\x00\x00\x00\x00\
\x00\x00\x00\x0e\x00\x01\x00\x00\x00\x01\x00\x00\x00\x00\
\x00\x00\x01\x7a\xbe\xae\x3d\xec\
"

qt_version = [int(v) for v in QtCore.qVersion().split('.')]
if qt_version < [5, 8, 0]:
    rcc_version = 1
    qt_resource_struct = qt_resource_struct_v1
else:
    rcc_version = 2
    qt_resource_struct = qt_resource_struct_v2

def qInitResources():
    QtCore.qRegisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

def qCleanupResources():
    QtCore.qUnregisterResourceData(rcc_version, qt_resource_struct, qt_resource_name, qt_resource_data)

qInitResources()


functions.py
from lxml import etree
import requests
import re
import os

headers = {
    'Referer': 'https://www.bilibili.com/',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36'
}

class CIDNotFoundException(Exception):
    pass

def qualityid(quality_name):
    if quality_name == '1080P':
        return 80
    elif quality_name == '720P':
        return 64
    elif quality_name == '480P':
        return 32
    elif quality_name == '360P':
        return 16

# 正则提取bv号
def bvre(bv):
    bv += '?'
    rx = '(BV[\S]*?)\\?'
    r_bv = re.findall(rx, bv)[0]
    return r_bv.replace('/', '')

# 获取视频名称
def name(bv, headers):
    return etree.HTML(requests.get('https://www.bilibili.com/video/' + bv, headers=headers).text).xpath('//*[@id="viewbox_report"]/h1/span/text()')[0]


# 获取分页信息以及对应的cid
def get_cid(bv, page_num):
    url = 'https://api.bilibili.com/x/web-interface/view/detail'
    param = {
        'bvid': '%s' % bv,
    }
    text = requests.get(url, params=param, headers=headers).json()
    text_cid = text['data']['View']['pages']
    for page in text_cid:
        if page['page'] == page_num:
            return page['cid']
    raise CIDNotFoundException('目标视频似乎没有第 ' + page_num + ' 页。')


# 获取视频url
def play_url(cid, bv, headers, quality):
    url = 'https://api.bilibili.com/x/player/playurl'
    param = {
        'cid': '%s' % cid,
        'bvid': '%s' % bv,
        'qn': '%s' % quality,
    }
    return requests.get(url, params=param, headers=headers).json()


# 请求视频并保存
def get_flv(name, bv, flv_url, headers, page_num):
    response = requests.get(flv_url, headers=headers)
    code = response.status_code
    text = response.content
    if code == 200:
       # 删除标题中的 / 等字符，防止文件路径错误
        name = name.replace('/', '').replace('|', '')
        filename = "./Downloads/%s - PAGE %d.flv" % (name, page_num)
        with open(filename, 'wb') as fp1:
            fp1.write(text)
        with open("./DownloadHistory.txt", 'a') as fp2:
            fp2.write(name+"\nhttps://www.bilibili.com/video/" + bv + "?p=%d\n\n" % page_num)
        return os.path.abspath(filename)
    else:
        response.raise_for_status()


ui.py
# -*- coding: utf-8 -*-

################################################################################
## Form generated from reading UI file 'ui.ui'
##
## Created by: Qt User Interface Compiler version 5.14.1
##
## WARNING! All changes made in this file will be lost when recompiling UI file!
################################################################################

from PySide2.QtCore import (QCoreApplication, QMetaObject, QObject, QPoint,
    QRect, QSize, QUrl, Qt)
from PySide2.QtGui import (QBrush, QColor, QConicalGradient, QCursor, QFont,
    QFontDatabase, QIcon, QLinearGradient, QPalette, QPainter, QPixmap,
    QRadialGradient)
from PySide2.QtWidgets import *

import icon_rc

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        if MainWindow.objectName():
            MainWindow.setObjectName(u"MainWindow")
        MainWindow.setEnabled(True)
        MainWindow.resize(460, 150)
        icon = QIcon()
        icon.addFile(u":/icon/icon.ico", QSize(), QIcon.Normal, QIcon.Off)
        MainWindow.setWindowIcon(icon)
        self.centralwidget = QWidget(MainWindow)
        self.centralwidget.setObjectName(u"centralwidget")
        self.download = QPushButton(self.centralwidget)
        self.download.setObjectName(u"download")
        self.download.setEnabled(True)
        self.download.setGeometry(QRect(340, 60, 71, 23))
        self.download.setCheckable(False)
        self.download.setAutoDefault(False)
        self.lineEdit = QLineEdit(self.centralwidget)
        self.lineEdit.setObjectName(u"lineEdit")
        self.lineEdit.setGeometry(QRect(40, 30, 381, 20))
        self.label = QLabel(self.centralwidget)
        self.label.setObjectName(u"label")
        self.label.setGeometry(QRect(105, 10, 271, 16))
        self.label_2 = QLabel(self.centralwidget)
        self.label_2.setObjectName(u"label_2")
        self.label_2.setGeometry(QRect(40, 63, 61, 16))
        self.quality = QComboBox(self.centralwidget)
        self.quality.addItem("")
        self.quality.addItem("")
        self.quality.addItem("")
        self.quality.addItem("")
        self.quality.setObjectName(u"quality")
        self.quality.setGeometry(QRect(85, 60, 69, 22))
        self.saveAtButton = QPushButton(self.centralwidget)
        self.saveAtButton.setObjectName(u"saveAtButton")
        self.saveAtButton.setGeometry(QRect(260, 60, 75, 23))
        self.label_3 = QLabel(self.centralwidget)
        self.label_3.setObjectName(u"label_3")
        self.label_3.setGeometry(QRect(10, 130, 331, 16))
        self.label_4 = QLabel(self.centralwidget)
        self.label_4.setObjectName(u"label_4")
        self.label_4.setGeometry(QRect(360, 130, 91, 16))
        self.label_6 = QLabel(self.centralwidget)
        self.label_6.setObjectName(u"label_6")
        self.label_6.setGeometry(QRect(40, 90, 31, 16))
        self.videoName = QLabel(self.centralwidget)
        self.videoName.setObjectName(u"videoName")
        self.videoName.setGeometry(QRect(10, 110, 441, 20))
        self.videoName.setAlignment(Qt.AlignCenter)
        self.videoPage = QSpinBox(self.centralwidget)
        self.videoPage.setObjectName(u"videoPage")
        self.videoPage.setGeometry(QRect(70, 87, 42, 22))
        self.videoPage.setMinimum(1)
        self.videoPage.setMaximum(999999999)
        self.saveOriginalVideo = QCheckBox(self.centralwidget)
        self.saveOriginalVideo.setObjectName(u"saveOriginalVideo")
        self.saveOriginalVideo.setGeometry(QRect(320, 90, 91, 16))
        self.videoFormatType = QComboBox(self.centralwidget)
        self.videoFormatType.addItem("")
        self.videoFormatType.addItem("")
        self.videoFormatType.addItem("")
        self.videoFormatType.setObjectName(u"videoFormatType")
        self.videoFormatType.setGeometry(QRect(225, 87, 51, 22))
        self.label_5 = QLabel(self.centralwidget)
        self.label_5.setObjectName(u"label_5")
        self.label_5.setGeometry(QRect(145, 90, 81, 16))
        MainWindow.setCentralWidget(self.centralwidget)

        self.retranslateUi(MainWindow)

        self.download.setDefault(True)


        QMetaObject.connectSlotsByName(MainWindow)
    # setupUi

    def retranslateUi(self, MainWindow):
        MainWindow.setWindowTitle(QCoreApplication.translate("MainWindow", u"Bilibili Downloader", None))
        self.download.setText(QCoreApplication.translate("MainWindow", u"\u4e0b\u8f7d!", None))
        self.label.setText(QCoreApplication.translate("MainWindow", u"\u5728\u4e0b\u9762\u7684\u6846\u91cc\u8f93\u5165 BV \u53f7\u6216\u8005\u5b8c\u6574\u7684B\u7ad9\u89c6\u9891\u94fe\u63a5:", None))
        self.label_2.setText(QCoreApplication.translate("MainWindow", u"\u6e05\u6670\u5ea6:", None))
        self.quality.setItemText(0, QCoreApplication.translate("MainWindow", u"1080P", None))
        self.quality.setItemText(1, QCoreApplication.translate("MainWindow", u"720P", None))
        self.quality.setItemText(2, QCoreApplication.translate("MainWindow", u"480P", None))
        self.quality.setItemText(3, QCoreApplication.translate("MainWindow", u"360P", None))

        self.saveAtButton.setText(QCoreApplication.translate("MainWindow", u"\u4fdd\u5b58\u5230...", None))
        self.label_3.setText(QCoreApplication.translate("MainWindow", u"\u63d0\u793a: \u4e0b\u8f7d\u7684\u89c6\u9891\u9ed8\u8ba4\u4fdd\u5b58\u5230\u5f53\u524d\u76ee\u5f55\u4e0b\u7684 \"Downloads\" \u76ee\u5f55", None))
        self.label_4.setText(QCoreApplication.translate("MainWindow", u"by SNWCreations", None))
        self.label_6.setText(QCoreApplication.translate("MainWindow", u"\u9875\u53f7:", None))
        self.videoName.setText(QCoreApplication.translate("MainWindow", u"\u8f93\u5165\u5b8c BV \u53f7\u540e, \u5728\u8f93\u5165\u6846\u91cc\u6309\u4e0b\u56de\u8f66\u5373\u53ef\u5728\u8fd9\u91cc\u770b\u5230\u89c6\u9891\u540d\u79f0", None))
        self.saveOriginalVideo.setText(QCoreApplication.translate("MainWindow", u"\u4fdd\u7559\u539f\u59cb\u89c6\u9891", None))
        self.videoFormatType.setItemText(0, QCoreApplication.translate("MainWindow", u"MP4", None))
        self.videoFormatType.setItemText(1, QCoreApplication.translate("MainWindow", u"MP3", None))
        self.videoFormatType.setItemText(2, QCoreApplication.translate("MainWindow", u"FLV", None))

        self.label_5.setText(QCoreApplication.translate("MainWindow", u"\u9009\u62e9\u6587\u4ef6\u683c\u5f0f:", None))
    # retranslateUi

